File: MovieHostingLinkGripper/src/main/java/com/sheryv/tools/movielinkgripper/Series.java
package com.sheryv.tools.movielinkgripper;

import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonProperty;
import lombok.Data;

import java.util.List;

@Data
public class Series {
    private final String name;
    private final int season;
    private final String lang;
    private final List<Episode> episodes;

    @JsonCreator(mode = JsonCreator.Mode.PROPERTIES)
    public Series(@JsonProperty("name") String name, @JsonProperty("season") int season, @JsonProperty("lang") String lang, @JsonProperty("episodes") List<Episode> episodes) {
        this.name = name;
        this.season = season;
        this.lang = lang;
        this.episodes = episodes;
    }
}

File: WebCrawler/src/main/kotlin/com/sheryv/tools/webcrawler/process/impl/streamingwebsite/common/model/Series.kt
package com.sheryv.tools.webcrawler.process.impl.streamingwebsite.common.model

import com.sheryv.util.io.FileUtils
import java.nio.file.Path
import java.time.LocalDate

data class Series(
  val id: Long,
  val title: String,
  val season: Int,
  val lang: String,
  val seriesUrl: String,
  val posterUrl: String? = null,
  val tvdbId: String? = null,
  val imdbId: String? = null,
  val releaseDate: LocalDate? = null,
  val episodes: List<Episode> = emptyList(),
) {
  
  fun generateDirectoryPathForSeason(): Path {
    return Path.of(FileUtils.fixFileNameWithColonSupport(title), String.format("Season %02d", season))
  }
  
  fun formattedString(): String {
    var output = """Series title: $title
      |Season: $season, ID: $id
      |
      |Episodes: ${episodes.size}
      |
    """.trimMargin()
    
    output += episodes.joinToString("\n") {
      "%2d. %-40s | %7d | %s".format(it.number, it.title, it.id, it.downloadUrl.toString())
    }
    return output
  }
}
